[SelectionDAG] Eliminate redundant setcc on comparison results

SavchenkoValeriy opened the pull request:
When comparisons produce all-zeros or all-ones in scalars or per lane in vectors, comparing results of such comparisons against 0 is an identity operation. This change eliminates redundant comparison instructions after another comparison operation.

SavchenkoValeriy committed:
[SelectionDAG] Eliminate redundant setcc on vector comparison results

For values with all lanes being either 0 or -1, comparing < 0 is an identity
operation.

davemgreen reviewed (commented):
It sounds like you are trying to reimplement computeKnownBits / computeNumSignBits and perform optimizations based on it. A lot of the test cases already simplify in IR, but maybe not through a shuffle. I didn't look too much into the details, what would we be missing to make this happen through computeNumSignBits instead? Is there a generic DAG combine that could handle it if more AArch64 nodes were producing results?

SavchenkoValeriy commented:
@davemgreen I moved it to SelectionDAG and predicated it on boolean result === -1/0 mask using `computeNumSignBits`. Thanks for the suggestion!

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -4748,6 +4748,27 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
+    if (Cond == ISD::SETLT && isNullOrNullSplat(N1))
+      Candidate = N0;
+    // setcc 0, RHS, setgt (equivalent to RHS < 0)
+    else if (Cond == ISD::SETGT && isNullOrNullSplat(N0))
```
Does the constant on the RHS come up anywhere, or is the zero usually canonicalised to the RHS?

SavchenkoValeriy replied:
Ah, I bet it gets canonicalized away. I'll remove the case

RKSimon commented:
I'm travelling right now, but does this overlap with https://github.com/llvm/llvm-project/pull/164946 ?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -4746,6 +4746,19 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
     }
   }
 
+  // setcc X, 0, setlt --> X  (when X is all sign bits)
```
Can you move them to be next to each other

DaKnig commented:
it should be also true across other value-preserving nodes:
- select
- min/max (signed/unsigned) (other such select-like nodes)
- isd::abdu
- bitwise: or/xor/and

DaKnig commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -4746,6 +4746,19 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
+  //
+  // When we know that X has 0 or -1 in each lane, this comparison will produce
+  // X. This is only true when boolean contents are represented via 0s and -1s.
+  if (OpVT.isVector() && VT == OpVT &&
```
why only when `isVector`?

SavchenkoValeriy replied:
I dropped `isVector` check, but overall I can't remove `VT == OpVT` because of the transformation logic. I found and added a case for architecture where `i1` is illegal and there we can trigger this fold.

DaKnig commented on llvm/test/CodeGen/AArch64/setcc-redundant-cmlt.ll:
```diff
@@ -0,0 +1,118 @@
+  %lt0 = icmp slt <4 x i32> %bc, zeroinitializer
+  %sel = select <4 x i1> %lt0, <4 x i32> %x, <4 x i32> %y
+  ret <4 x i32> %sel
+}
```
please add a test for `(abdu (setcc ...) (setcc ...))` even if its just to track the performance of signbits. same for `max` and `min` between them.

SavchenkoValeriy replied:
Interesting enough, `ComputeNumSignBits` doesn't support `ISD::ABDU`.

SavchenkoValeriy committed:
[SelectionDAG] Extend redundant setcc elimination to SETNE

For values with all lanes being either 0 or -1, comparing \!= 0 is also an
identity operation.

SavchenkoValeriy committed:
[AArch64][NFC] Add tests for sign bits propagation through various ops

SavchenkoValeriy committed:
[SelectionDAG] Support scalar case for redundant setcc

SavchenkoValeriy committed:
[AArch64][NFC] Update the test to check GIsel as well

SavchenkoValeriy committed:
[AArch64][NFC] Add test case for UGT 0

RKSimon reviewed (approved):
LGTM - please update the title + summary as they still say this is vector only
